packages/playwright-core/src/server/injected/clock.ts

```diff
@@ -53,26 +52,35 @@ type Timer = {
 };
 
 interface Embedder {
-  postTask(task: () => void): void;
+  dateNow(): number;
+  performanceNow(): DOMHighResTimeStamp;
+  postTask(task: () => void, timeout?: number): () => void;
   postTaskPeriodically(task: () => void, delay: number): () => void;
 }
 
+type Time = {
+  time: number;
+  ticks: number;
+  frozen: boolean;
+  origin: number;
+};
+
+type LogEntryType = 'fastForward' | 'fastForwardTo' | 'install' | 'pause' | 'resume' | 'runFor' | 'setFixedTime' | 'setSystemTime';
+
 export class ClockController {
-  readonly timeOrigin: number;
-  private _now: { time: number, ticks: number, timeFrozen: boolean };
-  private _loopLimit: number;
+  readonly _now: Time;
   private _duringTick = false;
   private _timers = new Map<number, Timer>();
   private _uniqueTimerId = idCounterStart;
   private _embedder: Embedder;
   readonly disposables: (() => void)[] = [];
+  private _realTime: { startTicks: number, lastSyncTicks: number } | undefined;
+  private _log: { type: LogEntryType, time: number, param?: number }[] = [];
+  private _currentRealTimeTimer: { callAt: number, dispose: () => void } | undefined;
```

Comment:
Make `Date.now()` aware of "time is ticking" and recalculate on every get operation. Otherwise, it will stall without any scheduled timers, even though the page is active.